### PR TITLE
Add a build-script-impl option --build-sil-debugging-stdlib to compil…

### PR DESCRIPTION
**Pull request by eeckstein**
#### Resolved bug number: ([SR-](https://bugs.swift.org/browse/SR-))

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

 **Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| OS X platform | @swift-ci Please test OS X platform |
| Linux platform | @swift-ci Please test Linux platform |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

…e the stdlib with SIL debug info (-gsil)

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci Please smoke test

File: utils/build-script-impl [hunk 1]
```diff
@@ -158,6 +158,7 @@ KNOWN_SETTINGS=(
     build-swift-static-stdlib   "0"              "set to 1 to build static versions of the Swift standard library and SDK overlay"
     build-swift-examples        "1"              "set to 1 to build examples"
     build-serialized-stdlib-unittest "0"         "set to 1 to build the StdlibUnittest module with -sil-serialize-all"
+    build-sil-debugging-stdlib  "0"              "set to 1 to build the Swift standard library with -gsil to enable debugging and profiling on SIL level"
```
Comment (modocache) [inline utils/build-script-impl]: I believe the `0` here is redundant--couldn't this simply be the empty string `""`?

**Comment by modocache** [comment]
[SR-237](https://bugs.swift.org/browse/SR-237) tracks migrating the logic in `build-script-impl`, implemented as a shellscript, over to Python. This task is actually pretty difficult, in part because the shellscript is growing over time...

Would it make sense to allow this setting to be set via Python, rather than as a `build-script-impl`-only parameter? I'm not actually sure this is better, just curious if anyone has any thoughts here. @gribozavr?

File: utils/build-script-impl (hunk 1 above)
Comment (eeckstein) [inline utils/build-script-impl]: Frankly speaking, I don't know. I just copy-pasted from other options.

**Comment by gottesmm** [comment]
@modocache I agree with you. If it is deprecated, it should not be added to. I think the issue is that we have not made an effort to make it clear/easy on how to add an option such as this. I imagine an example that adds a simple argument like this would be sufficient. Thoughts?

**Comment by gribozavr** [comment]
While I agree that we should try not to add to the shell script, it does not seem to me that these kinds of customizations are not readily possible in the Python part.

**Comment by gottesmm** [comment]
@gribozavr If we are going to get rid of the shell script, we should figure out how to make these changes in the python part.

Otherwise, we will never get rid of build-script-impl. If that is our goal, we should not shirk.

**Comment by gribozavr** [comment]
@gottesmm Should this block progress though?  Adding all the necessary plumbing to the python script is a lot of work.

**Comment by gottesmm** [comment]
@gribozavr I am not saying that it should block progress. I am saying that we should figure out now what we need to do to get this missing functionality into build-script. I would like to get to the future where saying that adding to build-script-impl is not acceptable does not block progress.

**Comment by gribozavr** [comment]
Well, that's what https://bugs.swift.org/browse/SR-237 is tracking.  I believe it is a lot of work.  Adding plumbing for just one option that generates CMake arguments will get us 90% there, and then porting options one by one is mostly trivial.  But I believe that it will actually involve moving calls to cmake from `build-script-impl` to `build-script` :(

**Comment by modocache** [comment]
This all makes sense to me! Thanks for weighing in.

> If we are going to get rid of the shell script, we should figure out how to make these changes in the python part.

I've been trying to move things over piece by piece, although it might be worth rethinking that approach. Please leave feedback on [SR-237](https://bugs.swift.org/browse/SR-237)!

**Pull request merged**